crates/kv-store/src/sstable.rs

```diff
@@ -0,0 +1,1056 @@
+                if end_idx == table_idx {
+                    iter.back_to_key(end);
+                    // if the next back is invalid, the next should also be invalid
+                    if !iter.has_next_back() {
+                        iter.next();
+                    }
+                    (end_idx, None, None)
+                } else {
+                    let block = table.read_block_cached(end_idx);
+                    let iter = BlockIter::new_back_to_key(block, end);
+                    (end_idx, Some(iter), None)
+                }
+            }
+            Bound::Excluded(end) => {
+                let end_idx = table.find_back_block_idx(end);
+                if end_idx == table_idx {
+                    iter.back_to_key(end);
+                    // if the next back is invalid, the next should also be invalid
+                    if !iter.has_next_back() {
+                        iter.next();
+                    }
+                    (end_idx, None, Some(end))
+                } else {
+                    let block = table.read_block_cached(end_idx);
+                    let iter = BlockIter::new_back_to_key(block, end);
+                    (end_idx, Some(iter), Some(end))
+                }
+            }
+            Bound::Unbounded => {
+                let end_idx = table.meta.len() - 1;
+                if end_idx == table_idx {
+                    (end_idx, None, None)
+                } else {
+                    let block = table.read_block_cached(end_idx);
+                    let iter = BlockIter::new(block);
+                    (end_idx, Some(iter), None)
+                }
+            }
+        };
+
+        let mut ans = if let Some(end_iter) = end_iter {
+            debug_assert!(end_idx > table_idx);
+            SsTableIter {
+                table,
+                iter: SsTableIterInner::Double {
+                    front: iter,
+                    back: end_iter,
+                },
+                next_block_idx: table_idx,
+                back_block_idx: end_idx as isize,
+            }
+        } else {
+            debug_assert!(end_idx == table_idx);
+            SsTableIter {
+                table,
+                iter: SsTableIterInner::Same(iter),
+                next_block_idx: table_idx,
+                back_block_idx: end_idx as isize,
+            }
+        };
+        // the current iter may be empty, but has next iter. we need to skip the empty iter
+        ans.skip_next_empty();
+        ans.skip_next_back_empty();
+
+        if let Some(key) = excluded {
+            if ans.has_next() && ans.peek_next_key().unwrap() == key {
+                ans.next();
+            }
+        }
+        if let Some(key) = end_excluded {
+            if ans.has_next_back() && ans.peek_next_back_key().unwrap() == key {
+                ans.next_back();
+            }
+        }
+
+        // need to skip empty block
+        ans.skip_next_empty();
```

Comment:
We shouldn't need these two lines right? It's handled by .next() and .next_back() already

        ans.skip_next_empty();
        ans.skip_next_back_empty();

Reply:
yes